kv/kvserver: TestReplicaRangefeedRetryErrors failed

[(kv/kvserver).TestReplicaRangefeedRetryErrors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2189976&tab=buildLog) on [master@ab148da515e0db2f39219c7337ea167a0fa0eb6a](https://github.com/cockroachdb/cockroach/commits/ab148da515e0db2f39219c7337ea167a0fa0eb6a):

```
=== RUN   TestReplicaRangefeedRetryErrors
--- FAIL: TestReplicaRangefeedRetryErrors (17.33s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors770708885
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors770708885
    --- FAIL: TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING (3.24s)
        replica_rangefeed_test.go:721: incorrect events on stream, found [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> >  checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<wall_time:1597717820509037952 > > ], want [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> > ]

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaRangefeedRetryErrors PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaRangefeedRetryErrors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestReplicaRangefeedRetryErrors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2191151&tab=buildLog) on [master@22a31b641abbb5ec59023eb73dde63146159b515](https://github.com/cockroachdb/cockroach/commits/22a31b641abbb5ec59023eb73dde63146159b515):

```
=== RUN   TestReplicaRangefeedRetryErrors
--- FAIL: TestReplicaRangefeedRetryErrors (19.80s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors509800568
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors509800568
    --- FAIL: TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING (4.11s)
        replica_rangefeed_test.go:721: incorrect events on stream, found [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> >  checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<wall_time:1597731230952485514 > > ], want [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> > ]

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestReplicaRangefeedRetryErrors PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaRangefeedRetryErrors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestReplicaRangefeedRetryErrors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2190545&tab=buildLog) on [master@cb61c0374d15d409621bb7e811e517116ff6e1e1](https://github.com/cockroachdb/cockroach/commits/cb61c0374d15d409621bb7e811e517116ff6e1e1):

```
=== RUN   TestReplicaRangefeedRetryErrors
--- FAIL: TestReplicaRangefeedRetryErrors (14.21s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors307526741
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors307526741
    --- FAIL: TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING (2.90s)
        replica_rangefeed_test.go:721: incorrect events on stream, found [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> >  checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<wall_time:1597730454436364973 > > ], want [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> > ]

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaRangefeedRetryErrors PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaRangefeedRetryErrors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestReplicaRangefeedRetryErrors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2192539&tab=buildLog) on [master@22a31b641abbb5ec59023eb73dde63146159b515](https://github.com/cockroachdb/cockroach/commits/22a31b641abbb5ec59023eb73dde63146159b515):

```
=== RUN   TestReplicaRangefeedRetryErrors
--- FAIL: TestReplicaRangefeedRetryErrors (16.68s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors494881773
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors494881773
    --- FAIL: TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING (3.22s)
        replica_rangefeed_test.go:721: incorrect events on stream, found [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> >  checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<wall_time:1597742832436076685 > > ], want [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> > ]

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaRangefeedRetryErrors PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaRangefeedRetryErrors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestReplicaRangefeedRetryErrors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2192652&tab=buildLog) on [master@e66e7870f4007e3769f8d34a57d3740bdce4163b](https://github.com/cockroachdb/cockroach/commits/e66e7870f4007e3769f8d34a57d3740bdce4163b):

```
=== RUN   TestReplicaRangefeedRetryErrors
--- FAIL: TestReplicaRangefeedRetryErrors (17.05s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors572115819
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors572115819
    --- FAIL: TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING (3.21s)
        replica_rangefeed_test.go:721: incorrect events on stream, found [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> >  checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<wall_time:1597751214570927754 > > ], want [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> > ]

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaRangefeedRetryErrors PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaRangefeedRetryErrors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestReplicaRangefeedRetryErrors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2192803&tab=buildLog) on [master@b82ab8f3412dd2e1367c41bce15651908be5c9c1](https://github.com/cockroachdb/cockroach/commits/b82ab8f3412dd2e1367c41bce15651908be5c9c1):

```
=== RUN   TestReplicaRangefeedRetryErrors
--- FAIL: TestReplicaRangefeedRetryErrors (15.91s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors544000229
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors544000229
    --- FAIL: TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING (2.95s)
        replica_rangefeed_test.go:721: incorrect events on stream, found [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> >  checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<wall_time:1597763087850580099 > > ], want [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> > ]

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaRangefeedRetryErrors PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaRangefeedRetryErrors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestReplicaRangefeedRetryErrors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2193252&tab=buildLog) on [master@43f737f0e0aabf17775ec40b3a30388089e19b89](https://github.com/cockroachdb/cockroach/commits/43f737f0e0aabf17775ec40b3a30388089e19b89):

```
=== RUN   TestReplicaRangefeedRetryErrors
--- FAIL: TestReplicaRangefeedRetryErrors (16.44s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors773797114
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors773797114
    --- FAIL: TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING (3.11s)
        replica_rangefeed_test.go:721: incorrect events on stream, found [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> >  checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<wall_time:1597768553685727639 > > ], want [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> > ]

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaRangefeedRetryErrors PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaRangefeedRetryErrors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver).TestReplicaRangefeedRetryErrors failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2193468&tab=buildLog) on [master@4de8cbe500576567017a8adda6a222bb2eed58f5](https://github.com/cockroachdb/cockroach/commits/4de8cbe500576567017a8adda6a222bb2eed58f5):

```
=== RUN   TestReplicaRangefeedRetryErrors
--- FAIL: TestReplicaRangefeedRetryErrors (16.08s)
    test_log_scope.go:154: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors689787938
    test_log_scope.go:63: use -show-logs to present logs inline
=== RUN   TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestReplicaRangefeedRetryErrors689787938
    --- FAIL: TestReplicaRangefeedRetryErrors/REASON_LOGICAL_OPS_MISSING (2.99s)
        replica_rangefeed_test.go:721: incorrect events on stream, found [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> >  checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<wall_time:1597770111033005027 > > ], want [checkpoint:<span:<key:"\272" end_key:"\372" > resolved_ts:<> > ]

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestReplicaRangefeedRetryErrors PKG=./pkg/kv/kvserver TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestReplicaRangefeedRetryErrors.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/replica_rangefeed_test.go
+++ b/pkg/kv/kvserver/replica_rangefeed_test.go
@@ -24,7 +24,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
@@ -357,8 +356,6 @@ func TestReplicaRangefeedRetryErrors(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.WithIssue(t, 52966)
-
 	ctx := context.Background()
 	startKey := []byte("a")
 
@@ -395,11 +392,16 @@ func TestReplicaRangefeedRetryErrors(t *testing.T) {
 		subT *testing.T, stream *testStream, streamErrC <-chan *roachpb.Error, span roachpb.Span,
 	) {
 		subT.Helper()
-		expEvents := []*roachpb.RangeFeedEvent{
-			{Checkpoint: &roachpb.RangeFeedCheckpoint{
+		noResolveTimestampEvent := roachpb.RangeFeedEvent{
+			Checkpoint: &roachpb.RangeFeedCheckpoint{
 				Span:       span,
 				ResolvedTS: hlc.Timestamp{},
-			}},
+			},
+		}
+		resolveTimestampEvent := roachpb.RangeFeedEvent{
+			Checkpoint: &roachpb.RangeFeedCheckpoint{
+				Span: span,
+			},
 		}
 		var events []*roachpb.RangeFeedEvent
 		testutils.SucceedsSoon(t, func() error {
@@ -407,9 +409,8 @@ func TestReplicaRangefeedRetryErrors(t *testing.T) {
 				// Break if the error channel is already populated.
 				return nil
 			}
-
 			events = stream.Events()
-			if len(events) < len(expEvents) {
+			if len(events) < 1 {
 				return errors.Errorf("too few events: %v", events)
 			}
 			return nil
@@ -417,9 +418,21 @@ func TestReplicaRangefeedRetryErrors(t *testing.T) {
 		if len(streamErrC) > 0 {
 			subT.Fatalf("unexpected error from stream: %v", <-streamErrC)
 		}
+		expEvents := []*roachpb.RangeFeedEvent{&noResolveTimestampEvent}
+		if len(events) > 1 {
+			// Unfortunately there is a timing issue here and the range feed may
+			// publish two checkpoints, one with a resolvedTs and one without, so we
+			// check for either case.
+			resolveTimestampEvent.Checkpoint.ResolvedTS = events[1].Checkpoint.ResolvedTS
+			expEvents = []*roachpb.RangeFeedEvent{
+				&noResolveTimestampEvent,
+				&resolveTimestampEvent,
+			}
+		}
 		if !reflect.DeepEqual(events, expEvents) {
 			subT.Fatalf("incorrect events on stream, found %v, want %v", events, expEvents)
 		}
+
 	}
 
 	assertRangefeedRetryErr := func(

